src/meshes/t8code_mesh.jl

```diff
@@ -807,42 +824,26 @@ function adapt_callback_wrapper(forest,
 end
 
 """
-    Trixi.adapt!(mesh::T8codeMesh, adapt_callback; kwargs...)
+    Trixi.adapt_forest!(forest::Ptr{t8_forest}, adapt_callback; kwargs...)
 
-Adapt a `T8codeMesh` according to a user-defined `adapt_callback`.
+Adapt a `T8codeMesh` according to a user-defined `adapt_callback`. This
+function is primarily for internal use. See `Trixi.adapt!(mesh::T8codeMesh,
+adapt_callback; kwargs...)` for a more detailed documentation.
 
 # Arguments
-- `mesh::T8codeMesh`: Initialized mesh object.
+- `forest::Ptr{t8_forest}`: New forest.
 - `adapt_callback`: A user-defined callback which tells the adaption routines
                     if an element should be refined, coarsened or stay unchanged.
+- `kwargs`: Refer to `Trixi.adapt!(mesh::T8codeMesh, adapt_callback; kwargs...)`.
 
-    The expected callback signature is as follows:
-
-      `adapt_callback(forest, ltreeid, eclass_scheme, lelemntid, elements, is_family, user_data)`
-        # Arguments
-        - `forest`: Pointer to the analyzed forest.
-        - `ltreeid`: Local index of the current tree where the analyzed elements are part of.
-        - `eclass_scheme`: Element class of `elements`.
-        - `lelemntid`: Local index of the first element in `elements`.
-        - `elements`: Array of elements. If consecutive elements form a family
-                      they are passed together, otherwise `elements` consists of just one element.
-        - `is_family`: Boolean signifying if `elements` represents a family or not.
-        - `user_data`: Void pointer to some arbitrary user data. Default value is `C_NULL`.
-        # Returns
-          -1 : Coarsen family of elements.
-           0 : Stay unchanged.
-           1 : Refine element.
+Note that the old forest usually gets deallocated within t8code. Call
+`t8_forest_ref(Ref(mesh.forest))` beforehand to prevent that.
 
-- `kwargs`:
-    - `recursive = true`: Adapt the forest recursively. If true the caller must ensure that the callback
-                          returns 0 for every analyzed element at some point to stop the recursion.
-    - `balance = true`: Make sure the adapted forest is 2^(NDIMS-1):1 balanced.
-    - `partition = true`: Partition the forest to redistribute elements evenly among MPI ranks.
-    - `ghost = true`: Create a ghost layer for MPI data exchange.
-    - `user_data = C_NULL`: Pointer to some arbitrary user-defined data.
+Returns a `Ptr{t8_forest}` to a new forest.
 """
-function adapt(forest::Ptr{t8_forest}, adapt_callback; recursive = true, balance = true,
-               partition = true, ghost = true, user_data = C_NULL)
+function adapt_forest(forest::Ptr{t8_forest}, adapt_callback; recursive = true,
```

Comment:
```suggestion
function adapt_forest(forest::Union{T8code.ForestWrapper, Ptr{t8_forest}}, adapt_callback; recursive = true,
```
See for `balance_forest`.